samples: doc: Added README for POSIX samples

dougwfoster opened the pull request:
Added README.rst for eventfd and uname POSIX samples.
Helps fix #27805

dougwfoster committed:
samples: doc: Added/Updated README and Makefile for POSIX samples

Added README.rst for eventfd and uname samples. Updated README for
gettimeofday to align with other READMEs.Updated Makefile.host
file for samples to store output file in 'build' directory.

Signed-off-by: Doug Foster <dougwfost@gmail.com>

cfriedt reviewed (dismissed):
LGTM - a great follow up commit or PR would be to list POSIX samples in the os services / portability / POSIX doc area.

Would be happy to re-approve if you wish to add that here, @dougwfoster ?

ycsin commented on samples/posix/eventfd/README.rst:
```diff
@@ -0,0 +1,46 @@
+.. _posix-eventfd-sample:
+
+POSIX eventfd()
+########################################################
```
IIRC, their length should match?

```suggestion
POSIX eventfd()
###############
```

ycsin commented on samples/posix/uname/README.rst:
```diff
@@ -0,0 +1,56 @@
+.. _posix-uname-sample:
+
+POSIX uname()
+########################################################
```
```suggestion
POSIX uname()
#############
```

kartben commented on samples/posix/uname/README.rst:
```diff
@@ -0,0 +1,56 @@
+Overview
+********
+
+In this sample application, the POSIX uname() function
```
adding an hyperlink to uname()'s doc would be welcome
https://pubs.opengroup.org/onlinepubs/9699919799/functions/uname.html

cfriedt replied:
@kartben - still on the backburner, but I think we will eventually add doxygen for all posix api calls, with links to the upstream specs. It's just such a big endeavour, that I would probably want to do it in an automated way.

kartben replied:
@cfriedt you're suggesting writing original docs, right? Because I don't think the Open Group license would allow us to reuse/redistribute their documentation?

kartben commented on samples/posix/eventfd/README.rst:
```diff
@@ -0,0 +1,46 @@
+Overview
+********
+
+This sample application demonstrates using the POSIX eventfd()
```
adding a reference to eventfd() doc would be welcome here, but not sure why there's none on opengroup webpage? @cfriedt might know

cfriedt replied:
`eventfd()` is not yet part of POSIX. It just happens to be lumped in with POSIX in Zephyr because it's available on most POSIX systems, is mostly dependent already on existing POSIX APIs, etc.

kartben commented on samples/posix/eventfd/README.rst:
```diff
@@ -0,0 +1,46 @@
+The returned file descriptor is used with write/read calls and write/read values
+are output to the console.
+
+
```
nit: remove extra line

```suggestion
```

kartben commented on samples/posix/uname/README.rst:
```diff
@@ -0,0 +1,56 @@
+the console. Additionally, uname is added as a shell command and system
+information is displayed according to the option(s) provided for the command.
+
+
```
nit: remove

```suggestion
```

kartben commented on samples/posix/uname/README.rst:
```diff
@@ -0,0 +1,56 @@
+    machine[4]: x86
+
+
+    uart:~$ uname -a
+    Zephyr zephyr 3.5.99 zephyr-v3.5.0-3515-g10156f5f1d9c Jan  9 2024 22:23:04 x86 qemu_x86
+    uart:~$ uname -s
+    Zephyr
+    uart:~$ uname -n
+    zephyr
+    uart:~$ uname -r
+    3.5.99
+    uart:~$ uname -v
+    zephyr-v3.5.0-3515-g10156f5f1d9c Jan  9 2024 22:23:04
+    uart:~$ uname -m
+    x86
+    uart:~$ uname -i
+    qemu_x86
+    uart:~$ uname -smi
+    Zephyr x86 qemu_x86
```
I don't think showing all the possible ways to run the sample is really necessary.
It certainly makes sense to have at least one with the multiple options though

```suggestion
    uart:~$ uname -a
    Zephyr zephyr 3.5.99 zephyr-v3.5.0-3515-g10156f5f1d9c Jan  9 2024 22:23:04 x86 qemu_x86
    uart:~$ uname -smi
    Zephyr x86 qemu_x86
```

kartben reviewed (changes requested):
Thanks!
Additional note: line wrapping is a bit inconsistent in the two proposed files. I realize this might be due to copy-pasting from the other POSIX sample README, but it would make sense to adjust to have consistent 100-character long lines.

dougwfoster commented:
> LGTM - a great follow up commit or PR would be to list POSIX samples in the os services / portability / POSIX doc area.
> 
> Would be happy to re-approve if you wish to add that here, @dougwfoster ?

That's a good idea. I can add a **POSIX Sample** section there or list them in the **POSIX Applications in Zephyr** section.

dougwfoster commented on samples/posix/uname/README.rst:
```diff
@@ -0,0 +1,56 @@
+   :board: qemu_x86
+   :goals: run
+   :compact:
+
```
I noticed the uname sample does not have a Makefile.host file. Is there a specific reason it doesn't have one? I will add one and see if it builds.

ycsin replied:
@dougwfoster this now can be done since #67648 is merged

dougwfoster commented on samples/posix/eventfd/README.rst:
```diff
@@ -0,0 +1,46 @@
+
+.. code-block:: console
+
+   make -f Makefile.host
```
Thanks for pointing bringing this to my attention. It looks like the Makefile.host file is actually set up to work only if the cwd is **samples/posix/eventfd**, so the command to build it would be `make -f Makefile.host`. I can update the Makefile.host files so it can be built with  `make -f samples/posix/eventfd/Makefile.host`, similar to how the west build would work, and update the instructions.

cfriedt commented:
@dougwfoster - I think the samples portion should probably be automatically built (as long as the .rst files are captured by the main sample rst)

From the os services / portability / POSIX page, maybe just a link back to the samples would be good?

@kartben could probably be more specific.

dougwfoster committed:
doc: posix: Update services/portability/posix doc

Update index.rst located in doc/services/portability/posix/overview
to include link to POSIX samples.

Signed-off-by: Doug Foster <dougwfost@gmail.com>

kartben commented on samples/posix/eventfd/README.rst:
```diff
@@ -0,0 +1,45 @@
+
+.. code-block:: console
+
+   make -f samples/posix/eventfd/Makefile.host
+
+The make output file will be located in samples/posix/eventfd.
```
non blocking: kind of not ideal that this adds a new untracked file to the git repo, but I guess it was already like that before.

ycsin commented on samples/posix/eventfd/README.rst:
```diff
@@ -0,0 +1,45 @@
+Building and Running
+********************
+
+This project outputs to the console.  It can be built and executed on QEMU as follows:
```
```suggestion
This project outputs to the console. It can be built and executed on QEMU as follows:
```

ycsin commented on samples/posix/uname/README.rst:
```diff
@@ -0,0 +1,43 @@
+Building and Running
+********************
+
+This project outputs to the console.  It can be built and executed on QEMU as follows:
```
```suggestion
This project outputs to the console. It can be built and executed on QEMU as follows:
```

ycsin commented on samples/posix/eventfd/Makefile.host:
```diff
@@ -1,4 +1,4 @@
 
-eventfd: src/main.c
-	$(CC) $^ -o $@
+eventfd: samples/posix/eventfd/src/main.c
```
> Updated Makefile.hostfile for samples to be used with make from 'zephyr' directory, similar to west build command.

I find this change somewhat unnecessary, it can't be compared with the `west build` command because `west` works anywhere (either in the `zephyr` folder or in the `zephyr/samples/posix/eventfd` folder). Changing the path in this Makefile doesn't fix its limitation, it's just a matter of preference (i.e. I usually build in the `zephyrproject` folder, this change wouldn't benefit me) and it deviates from other existing `Makefile.host`s in the `net/socket` examples.

Also, before this change, a developer can copy the sample to anywhere and `make` would work as-is, after this change, the sample must be in the relative path for the Makefile to work.

TL;DR: This change only increases the limitation of the Makefile

ycsin commented on samples/posix/gettimeofday/Makefile.host:
```diff
@@ -1,4 +1,4 @@
 
-gettimeofday: src/main.c
-	$(CC) $^ -o $@
+gettimeofday: samples/posix/gettimeofday/src/main.c
+	$(CC) $^ -o samples/posix/gettimeofday/$@
```
If we output the binary to a `build` folder, then it will be ignored by git and won't be an untracked file cc @kartben :

```suggestion
gettimeofday: src/main.c
	mkdir -p build
	$(CC) $^ -o build/$@
```

dougwfoster replied:
Thanks for the suggestion. I made this change in my latest update.

cfriedt reviewed (approved):
LGTM - the only slight inaccuracy is that events is not POSIX (yet), even though it is available on many POSIX operating systems.

Otherwise, I think maybe @kartben could potentially offer some style suggestions?
